ddsketch/store/buffered_paginated.go

```diff
@@ -0,0 +1,495 @@
+	} else {
+		// Extends s.pages right.
+		newPages := make([][]float64, s.newPagesLen(pageIndex-s.minPageIndex+1))
+		copy(newPages, s.pages)
+		s.pages = newPages
+	}
+
+	page := s.pages[pageIndex-s.minPageIndex]
+	if page == nil {
+		page = make([]float64, pageLen)
+		s.pages[pageIndex-s.minPageIndex] = page
+	}
+	return page
+}
+
+func (s *BufferedPaginatedStore) newPagesLen(required int) int {
+	// Grow in size by multiples of 64 bytes
+	pageGrowthIncrement := 64 * 8 / ptrSize
+	return (required + pageGrowthIncrement - 1) & -pageGrowthIncrement
+}
+
+// compact transfers indexes from the buffer to the pages. It only creates new
+// pages if they can encode enough buffered indexes so that it frees more space
+// in the buffer than the new page takes.
+func (s *BufferedPaginatedStore) compact() {
+	pageLen := 1 << s.pageLenLog2
+
+	s.sortBuffer()
+
+	for bufferPos := 0; bufferPos < len(s.buffer); {
+		bufferPageStart := bufferPos
+		pageIndex := s.pageIndex(s.buffer[bufferPageStart])
+		bufferPos++
+		for bufferPos < len(s.buffer) && s.pageIndex(s.buffer[bufferPos]) == pageIndex {
+			bufferPos++
+		}
+		bufferPageEnd := bufferPos
+
+		// We avoid creating a new page if it would take more memory space than
+		// what we would free in the buffer. Note that even when the page itself
+		// takes less memory space than the buffered indexes that can be encoded
+		// in the page, because we may have to extend s.pages, the store may end
+		// up larger. However, for the sake of simplicity, we ignore the length
+		// of s.pages.
+		ensureExists := (bufferPageEnd-bufferPageStart)*bufferEntrySize >= pageLen*float64size
+		newPage := s.page(pageIndex, ensureExists)
+		if newPage != nil {
+			for _, index := range s.buffer[bufferPageStart:bufferPageEnd] {
+				newPage[s.lineIndex(index)]++
+			}
+			copy(s.buffer[bufferPageStart:], s.buffer[bufferPageEnd:])
+			s.buffer = s.buffer[:len(s.buffer)+bufferPageStart-bufferPageEnd]
+			bufferPos = bufferPageStart
+		}
+	}
+
+	s.bufferCompactionTriggerLen = len(s.buffer) + pageLen
+}
+
+func (s *BufferedPaginatedStore) sortBuffer() {
+	sort.Slice(s.buffer, func(i, j int) bool { return s.buffer[i] < s.buffer[j] })
+}
+
+func (s *BufferedPaginatedStore) Add(index int) {
+	pageIndex := s.pageIndex(index)
+	if pageIndex >= s.minPageIndex && pageIndex < s.minPageIndex+len(s.pages) {
+		page := s.pages[pageIndex-s.minPageIndex]
+		if page != nil {
+			page[s.lineIndex(index)]++
+			return
+		}
+	}
+
+	// The page does not exist, use the buffer.
+	if len(s.buffer) == cap(s.buffer) && len(s.buffer) >= s.bufferCompactionTriggerLen {
+		s.compact()
+	}
+
+	s.buffer = append(s.buffer, index)
+}
+
+func (s *BufferedPaginatedStore) AddBin(bin Bin) {
+	s.AddWithCount(bin.Index(), bin.Count())
+}
+
+func (s *BufferedPaginatedStore) AddWithCount(index int, count float64) {
+	if count == 0 {
+		return
+	} else if count == 1 {
+		s.Add(index)
+	} else {
+		s.page(s.pageIndex(index), true)[s.lineIndex(index)] += float64(count)
```

Comment:
You don't need the `float64(count)` conversion.

Reply:
Good catch, fixed in cdc5d88a22f34cd85b1a1ce47d1b641fb16f62b3.